File: E2ETestResults/CoreSubset/Core/Utility/MiscUtil.ts
import { FeatureFlagIds, tsConfig } from 'TypeDefs';

import { FeatureFlags } from './FeatureFlags';

import { JsNativeExtensionMethods, TypeUtil } from 'NativeJsTypeDefs';

import 'mscorlib';

import { _ } from 'Underscore';

import { Utility } from '../../Bootstrap/Utility';

import { WindowHelper } from '../../CoreSlim/Utility/WindowHelper';

export enum PathnameKey {
  WorkbookName = 2,
  SheetId = 3,
  AuthoringSheet = 4,
}

/**
 * Contains utility methods for etc.
 */
export class MiscUtil {
  public static get_pathName(): string {
    let window: Object = Utility.locationWindow;
    return WindowHelper.getLocation(window).pathname;
  }

  public static get_urlPathnameParts(): { [key: number]: string } {
    let pathname: string = MiscUtil.pathName;
    let siteRoot: string = tsConfig.site_root;
    let index: number = pathname.indexOf(siteRoot, 0);
    let actualPath: string = pathname.substr(index + siteRoot.length);
    let pathnameParts: string[] = actualPath.split('/');
    let pathnameProps: { [key: number]: string } = {};
    pathnameProps[PathnameKey.WorkbookName] = pathnameParts[<number>PathnameKey.WorkbookName];
    pathnameProps[PathnameKey.SheetId] = pathnameParts[<number>PathnameKey.SheetId];
    pathnameProps[PathnameKey.AuthoringSheet] = pathnameParts[<number>PathnameKey.AuthoringSheet];
    return pathnameProps;
  }

  /**
   * Performs a shallow equals of two given objects.
   * @param valueA An object or `null`
   * @param valueB Another object or `null`
   * @returns `true` if all keys are `===` between the two objects.
   */
  public static shallowEquals(valueA: any, valueB: any): boolean {
    if (valueA === valueB) {
      return true;
    }
    if (valueA === null || valueB === null) {
      return false;
    }
    let keysA: string[] = any.keys(valueA);
    let keysB: string[] = any.keys(valueB);
    if (keysA.length !== keysB.length) {
      return false;
    }
    for (let i = 0; i < keysA.length; i++) {
      let key: string = keysA[i];
      if (!valueB.hasOwnProperty(key) || valueA[key] !== valueB[key]) {
        return false;
      }
    }
    return true;
  }

  /**
   * Percent encodes (also called URL encoding) the given string.  First the browser's encodeUriComponent()
   * is called on the string and then any characters not included in `unreservedChars` are
   * then encoded.  If you only want the browser standard behavior then use {@link String.EncodeUriComponent},
   * this method is only useful for encoding extra characters that the browser might not encode.
   * @param valueToEncode The string to be encoded
   * @param unreservedChars A dictionary of unreserved string keys (single characters).  Any character
   * not included as a key will be escaped.
   * @returns The encoded string
   */
  public static percentEncode(valueToEncode: string, unreservedChars: { [key: string]: string }): string {
    valueToEncode = string.encodeURIComponent(valueToEncode);
    if (ss.isNullOrUndefined(unreservedChars)) {
      return valueToEncode;
    }
    let sb: ss.StringBuilder = new ss.StringBuilder();
    let i: number = 0;
    while (i < valueToEncode.length) {
      let s: string = valueToEncode.substr(i, 1);
      if (s === '%') {
        sb.append(valueToEncode.substr(i, 3));
        i += 2;
      } else
        if (!ss.keyExists(unreservedChars, s)) {
          sb.append('%').append((<number>s.charCodeAt(0)).toString(16).toUpperCase());
        } else {
          sb.append(s);
        }
      i++;
    }
    return sb.toString();
  }

  /**
   * Encodes string based on string encoding defined in wgapp\app\helpers\link_help.rb:encode_id(name)
   * First, specific characters are url-encoded, and the resulting string is then completely url-encoded.
   * @param valueToEncode The string to be encoded
   * @returns The encoded string
   */
  public static encodeForWG(valueToEncode: string): string {
    let usernameValidChars: { [key: string]: string } = {};
    let addCodes: (char: number, char: number) => void = (from: number, to: number) => {
      for (let i = from; i <= to; i++) {
        let s: string = string.fromCharCode(i);
        usernameValidChars[s] = s;
      }
    };
    addCodes('a', 'z');
    addCodes('A', 'Z');
    addCodes('0', '9');
    addCodes('_', '_');
    addCodes('-', '-');
    valueToEncode = MiscUtil.percentEncode(valueToEncode, usernameValidChars);
    valueToEncode = MiscUtil.percentEncode(valueToEncode, null);
    return valueToEncode;
  }

  /**
   * Tests whether a single argument or at least one of a list of arguments is null OR
   * undefined.
   * @param args argument to be checked
   * @returns true or false
   */
  public static isNullOrUndefined(args: any[]): boolean {
    if (ss.isNullOrUndefined(args)) {
      return true;
    }
    for (let i = 0; i < args.length; i++) {
      if (ss.isNullOrUndefined(args[i])) {
        return true;
      }
    }
    return false;
  }

  /**
   * Tests whether an array or a string is either null, undefined, or empty
   * (length is 0). You can pass one or more arguments to the function and it will
   * return true if at least one of the arguments is null or empty.
   * @param args argument to be checked
   * @returns true or false
   */
  public static isNullOrEmpty(args: any): boolean {
    if (ss.isNullOrUndefined(args)) {
      return true;
    }
    let dict: { [key: string]: number } = args;
    if (ss.isValue(dict['length']) && dict['length'] === 0) {
      return true;
    }
    return false;
  }

  /**
   * Calling this on a null string would be either an error or vacuously impossible in real C#, but since this
   * gets compiled to a static utility method in Saltarelle, it gives us a one-line way to check this condition
   * rather than three-line 'if (string.isNullOrEmpty()) {}' conditionals everywhere.
   */
  public static isNullOrEmpty$1(s: string): boolean {
    return ss.isNullOrEmptyString(s);
  }

  /**
   * Overload of same IsNullOrEmpty for generic List objects
   * typeparam T The object type contained by the List
   */
  public static isNullOrEmpty$2<T>(list: T[]): boolean {
    return list === null || list.length === 0;
  }

  /**
   * Overload of same IsNullOrEmpty for generic JsArray objects
   * typeparam T The object type contained by the List
   */
  public static isNullOrEmpty$3<T>(array: T[]): boolean {
    return array === null || array.length === 0;
  }

  /**
   * check if the given index is valid in the given array
   * @param index index of the array
   * @param arr the array to check
   * @returns true of false
   */
  public static isValidIndex(index: number, arr: any[]): boolean {
    return index >= 0 && index < arr.length;
  }

  /**
   * Is given string a truthy value as defined by Tableau's query param truthy values?
   * @param value 
   * @param defaultIfMissing 
   * @returns truthiness, or defaultIfMissing if the string is null or empty
   */
  public static toBoolean(value: string, defaultIfMissing: boolean): boolean {
    let positiveRegex: RegExp = new RegExp('^(yes|y|true|t|1)$', 'i');
    if (MiscUtil.isNullOrEmpty$1(value)) {
      return defaultIfMissing;
    }
    let match: string[] = value.match(positiveRegex);
    return !MiscUtil.isNullOrEmpty(match);
  }

  /**
   * Determines whether or not the user has the ability to view the data tab
   * @returns True if the data tab should be enabled, false if not
   */
  public static shouldShowDataTab(): boolean {
    return (tsConfig.allow_add_new_datasource && FeatureFlags.isEnabled(FeatureFlagIds.DataToTheWeb) && (!tsConfig.is_mobile || (tsConfig.is_mobile && FeatureFlags.isEnabled(FeatureFlagIds.DataToTheWebMobile))));
  }

  /**
   * Replaces the query in the given URI with a new set of key/value pairs.
   * @param uri A uri
   * @param parameters The set of query paramters to use
   * @returns A uri with the query replaced(or appended)
   */
  public static replaceUriQueryParameters(uri: Object, parameters: { [key: string]: string[] }): Object {
    if (parameters.count === 0) {
      return uri;
    }
    let newQueryString: ss.StringBuilder = new ss.StringBuilder();
    let first: boolean = true;
    let appendSeparator: () => void = () => {
      newQueryString.append(first ? '?' : '&');
      first = false;
    };
    for (const key of _.keys(parameters)) {
      let vals: string[] = parameters[key];
      let keyEncoded: string = string.encodeURIComponent(key);
      if (ss.isNullOrUndefined(vals) || vals.length === 0) {
        appendSeparator();
        newQueryString.append(keyEncoded);
      } else {
        for (const value of vals) {
          appendSeparator();
          newQueryString.append(keyEncoded).append('=').append(string.encodeURIComponent(value));
        }
      }
    }
    let hash: string = '';
    let baseUri: string = '';
    if (uri.length > 0) {
      let indexOfQuery: number = (<string>uri).indexOf('?');
      let indexOfHash: number = (<string>uri).indexOf('#');
      let indexOfEnd: number = Math.min(indexOfQuery < 0 ? (<string>uri).length : indexOfQuery, indexOfHash < 0 ? (<string>uri).length : indexOfHash);
      baseUri = (<string>uri).substr(0, indexOfEnd);
      hash = indexOfHash < 0 ? '' : (<string>uri).substr(indexOfHash);
    }
    return baseUri + newQueryString + hash;
  }

  /**
   * We can find that some of our bools are actually strings, like "false" or "true". c# won't catch this
   * and makes us write a silly method to convert our bool to another bool.
   * @param value The bool that needs sanatized
   * @returns Null/undefined if the value passed is null. Otherse it will pass back a parsed bool
   */
  public static sanatizeBoolean(value: boolean): boolean {
    if (ss.isNullOrUndefined(value)) {
      return value;
    }
    return value.toString().toLowerCase() === 'true';
  }

  /**
   * Disposes the given object if non-null.
   * typeparam T The type of the class to dispose
   * @returns returns null
   */
  public static dispose<T>(d: T): T {
    if (ss.isValue(d)) {
      d.dispose();
    }
    return null;
  }

  /**
   * Disposes a list of disposables.
   * typeparam T The type of the class to dispose
   * @returns returns null
   */
  public static dispose$1<T>(d: T[]): T[] {
    if (ss.isValue(d)) {
      for (const v of d) {
        if (ss.isValue(v)) {
          v.dispose();
        }
      }
      ss.clear(d);
    }
    return null;
  }

  /**
   * Safely calls Window.ClearTimeout for the provided handle, which was previously created using Window.SetTimeout.
   * Returns null so that you can clear the timeout and assign in one statement, e.g.:
   * myTimeout = MiscUtil.ClearTimeout(myTimeout);
   */
  public static clearTimeout(handle: number | null): number | null {
    if (handle.hasValue) {
      window.clearTimeout(handle.value);
    }
    return null;
  }

  public static clearInterval(handle: number | null): number | null {
    if (handle.hasValue) {
      window.clearInterval(handle.value);
    }
    return null;
  }

  /**
   * Returns a deep clone of the given object
   */
  public static cloneObject(src: any): any {
    let objStr: string = JSON.stringify(src, (k, v) => {
      return v;
    });
    return JSON.parse(objStr);
  }
}

File: test/SaltarelleProjectTests/CoreSubset/Core/Utility/MiscUtil.cs
// ---------------------------------------------------------------------------------------------------------------------
// <copyright file="MiscUtil.cs" company="Tableau Software">
//   This file is the copyrighted property of Tableau Software and is protected by registered patents and other
//   applicable U.S. and international laws and regulations.
//
//   Unlicensed use of the contents of this file is prohibited. Please refer to the NOTICES.txt file for further details.
// </copyright>
// ---------------------------------------------------------------------------------------------------------------------

namespace Tableau.JavaScript.Vql.Core
{
    using System;
    using System.Collections.Generic;
    using System.Collections.TypedArrays;
    using System.Diagnostics.CodeAnalysis;
    using System.Html;
    using System.Runtime.CompilerServices;
    using System.Serialization;
    using System.Text;
    using System.Text.RegularExpressions;
    using Tableau.JavaScript.CoreSlim;
    using Tableau.JavaScript.Vql.Bootstrap;
    using Tableau.JavaScript.Vql.TypeDefs;
    using UnderscoreJs;

    [NumericValues]
    public enum PathnameKey
    {
        WorkbookName = 2,
        SheetId = 3,
        AuthoringSheet = 4,
    }

    /// <summary>
    /// Contains utility methods for etc.
    /// </summary>
    public static class MiscUtil
    {
        public static string PathName
        {
            get
            {
                WindowInstance window = Utility.LocationWindow;
                return WindowHelper.GetLocation(window).Pathname;
            }
        }

        public static JsDictionary<PathnameKey, string> UrlPathnameParts
        {
            get
            {
                string pathname = MiscUtil.PathName;
                string siteRoot = TsConfig.SiteRoot;
                int index = pathname.IndexOf(siteRoot, 0);
                string actualPath = pathname.Substr(index + siteRoot.Length);

                string[] pathnameParts = actualPath.Split("/");

                JsDictionary<PathnameKey, string> pathnameProps = new JsDictionary<PathnameKey, string>();

                // TODO: change this to iterate over Enum.GetValues when we upgrade to script sharp version that supports this.
                pathnameProps[PathnameKey.WorkbookName] = pathnameParts[(int)PathnameKey.WorkbookName];
                pathnameProps[PathnameKey.SheetId] = pathnameParts[(int)PathnameKey.SheetId];
                pathnameProps[PathnameKey.AuthoringSheet] = pathnameParts[(int)PathnameKey.AuthoringSheet];
                return pathnameProps;
            }
        }

        /// <summary>
        /// Performs a shallow equals of two given objects.
        /// </summary>
        /// <param name="valueA">An object or <c>null</c></param>
        /// <param name="valueB">Another object or <c>null</c></param>
        /// <returns><c>true</c> if all keys are <c>===</c> between the two objects.</returns>
        public static bool ShallowEquals(object valueA, object valueB)
        {
            if (valueA == valueB) { return true; }
            if (valueA == null || valueB == null) { return false; }

            var keysA = Keys(valueA);
            var keysB = Keys(valueB);
            if (keysA.Length != keysB.Length)
            {
                return false;
            }
            for (int i = 0; i < keysA.Length; i++)
            {
                var key = keysA[i];
                if (!HasOwnProperty(valueB, key) || TypeUtil.GetField<object>(valueA, key) != TypeUtil.GetField<object>(valueB, key))
                {
                    return false;
                }
            }
            return true;
        }

        /// <summary>
        /// Percent encodes (also called URL encoding) the given string.  First the browser's encodeUriComponent()
        /// is called on the string and then any characters not included in <c>unreservedChars</c> are
        /// then encoded.  If you only want the browser standard behavior then use <seealso cref="string.EncodeUriComponent"/>,
        /// this method is only useful for encoding extra characters that the browser might not encode.
        /// </summary>
        /// <param name="valueToEncode">The string to be encoded</param>
        /// <param name="unreservedChars">A dictionary of unreserved string keys (single characters).  Any character
        /// not included as a key will be escaped.</param>
        /// <returns>The encoded string</returns>
        public static string PercentEncode(string valueToEncode, JsDictionary<string, string> unreservedChars)
        {
            valueToEncode = string.EncodeUriComponent(valueToEncode);

            if (Script.IsNullOrUndefined(unreservedChars)) { return valueToEncode; }

            StringBuilder sb = new StringBuilder();
            int i = 0;

            while (i < valueToEncode.Length)
            {
                string s = valueToEncode.Substr(i, 1);
                if (s == "%")
                {
                    // browser handled encoding, append unchanged
                    sb.Append(valueToEncode.Substr(i, 3));
                    i += 2;
                }
                else if (!unreservedChars.ContainsKey(s))
                {
                    sb.Append("%").Append(((int)s.CharCodeAt(0)).ToString(16).ToUpperCase());
                }
                else
                {
                    sb.Append(s);
                }
                i++;
            }

            return sb.ToString();
        }

        /// <summary>
        /// Encodes string based on string encoding defined in wgapp\app\helpers\link_help.rb:encode_id(name)
        /// First, specific characters are url-encoded, and the resulting string is then completely url-encoded.
        /// </summary>
        /// <param name="valueToEncode">The string to be encoded</param>
        /// <returns>The encoded string</returns>
        public static string EncodeForWG(string valueToEncode)
        {
            JsDictionary<string, string> usernameValidChars = new JsDictionary<string, string>();

            Action<char, char> addCodes = (char from, char to) =>
            {
                for (char i = from; i <= to; i++)
                {
                    string s = string.FromCharCode(i);
                    usernameValidChars[s] = s;
                }
            };

            addCodes('a', 'z');
            addCodes('A', 'Z');
            addCodes('0', '9');
            addCodes('_', '_');
            addCodes('-', '-');

            valueToEncode = PercentEncode(valueToEncode, usernameValidChars);
            valueToEncode = PercentEncode(valueToEncode, null);
            return valueToEncode;
        }

        /// <summary>
        /// Tests whether a single argument or at least one of a list of arguments is null OR
        /// undefined.
        /// </summary>
        /// <param name="args">argument to be checked</param>
        /// <returns>true or false</returns>
        public static bool IsNullOrUndefined(object[] args)
        {
            if (Script.IsNullOrUndefined(args))
            {
                return true;
            }

            for (int i = 0; i < args.Length; i++)
            {
                if (Script.IsNullOrUndefined(args[i]))
                {
                    return true;
                }
            }

            return false;
        }

        /// <summary>
        /// Tests whether an array or a string is either null, undefined, or empty
        /// (length is 0). You can pass one or more arguments to the function and it will
        /// return true if at least one of the arguments is null or empty.
        /// </summary>
        /// <param name="args">argument to be checked</param>
        /// <returns>true or false</returns>
        public static bool IsNullOrEmpty(object args)
        {
            if (Script.IsNullOrUndefined(args))
            {
                return true;
            }

            // B46804: corrected semantics for this function.
            JsDictionary<string, int> dict = JsDictionary<string, int>.GetDictionary(args);
            if (Script.IsValue(dict["length"]) && dict["length"] == 0)
            {
                return true;
            }

            return false;
        }

        /// <summary>
        /// Calling this on a null string would be either an error or vacuously impossible in real C#, but since this
        /// gets compiled to a static utility method in Saltarelle, it gives us a one-line way to check this condition
        /// rather than three-line 'if (string.isNullOrEmpty()) {}' conditionals everywhere.
        /// </summary>
        public static bool IsNullOrEmpty(this string s)
        {
            return string.IsNullOrEmpty(s);
        }

        /// <summary>
        /// Overload of same IsNullOrEmpty for generic List objects
        /// <param name="list">the List object to check</param>
        /// </summary>
        /// <typeparam name="T">The object type contained by the List</typeparam>
        [IncludeGenericArguments(false)]
        public static bool IsNullOrEmpty<T>(this List<T> list)
        {
            return list == null || list.Count == 0;
        }

        /// <summary>
        /// Overload of same IsNullOrEmpty for generic JsArray objects
        /// <param name="array">the array object to check</param>
        /// </summary>
        /// <typeparam name="T">The object type contained by the List</typeparam>
        [IncludeGenericArguments(false)]
        public static bool IsNullOrEmpty<T>(this JsArray<T> array)
        {
            return array == null || array.Length == 0;
        }

        /// <summary>
        /// check if the given index is valid in the given array
        /// </summary>
        /// <param name="index">index of the array</param>
        /// <param name="arr">the array to check</param>
        /// <returns>true of false</returns>
        public static bool IsValidIndex(int index, Array arr)
        {
            return index >= 0 && index < arr.Length;
        }

        /// <summary>
        /// Checks if the given value is a non-null object.  Implementation cribbed from Underscore.
        /// </summary>
        /// <param name="o"></param>
        /// <returns>True if an object value</returns>
        [InlineCode("({o} === Object({o}))")]
        [SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", Justification = "Inline code")]
        public static bool IsObject(object o)
        {
            return false;
        }

        [InlineCode("{owner}.hasOwnProperty({field})")]
        [SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", Justification = "Inline code")]
        public static bool HasOwnProperty(object owner, string field)
        {
            return false;
        }

        /// <summary>
        /// Is given string a truthy value as defined by Tableau's query param truthy values?
        /// </summary>
        /// <param name="value"></param>
        /// <param name="defaultIfMissing"></param>
        /// <returns>truthiness, or defaultIfMissing if the string is null or empty</returns>
        public static bool ToBoolean(string value, bool defaultIfMissing)
        {
            var positiveRegex = new Regex(@"^(yes|y|true|t|1)$", "i");

            if (IsNullOrEmpty(value))
            {
                return defaultIfMissing;
            }

            string[] match = value.Match(positiveRegex);
            return !IsNullOrEmpty(match);
        }

        /// <summary>
        /// Determines whether or not the user has the ability to view the data tab
        /// </summary>
        /// <returns>True if the data tab should be enabled, false if not</returns>
        public static bool ShouldShowDataTab()
        {
            return (TsConfig.AllowAddNewDatasource && FeatureFlags.IsEnabled(FeatureFlagIds.DataToTheWeb) &&
                (!TsConfig.IsMobile || (TsConfig.IsMobile && FeatureFlags.IsEnabled(FeatureFlagIds.DataToTheWebMobile))));
        }

        /// <summary>
        /// Replaces the query in the given URI with a new set of key/value pairs.
        /// </summary>
        /// <param name="uri">A uri</param>
        /// <param name="parameters">The set of query paramters to use</param>
        /// <returns>A uri with the query replaced(or appended)</returns>
        public static URLStr ReplaceUriQueryParameters(URLStr uri, JsDictionary<string, JsArray<string>> parameters)
        {
            if (parameters.Count == 0)
            {
                return uri;
            }

            StringBuilder newQueryString = new StringBuilder();
            bool first = true;
            Action appendSeparator = delegate
            {
                newQueryString.Append(first ? "?" : "&");
                first = false;
            };

            foreach (string key in Underscore.Keys(parameters))
            {
                List<string> vals = parameters[key];
                string keyEncoded = string.EncodeUriComponent(key);
                if (Script.IsNullOrUndefined(vals) || vals.Count == 0)
                {
                    appendSeparator();
                    newQueryString.Append(keyEncoded);
                }
                else
                {
                    foreach (string value in vals)
                    {
                        appendSeparator();
                        newQueryString.Append(keyEncoded).Append("=").Append(string.EncodeUriComponent(value));
                    }
                }
            }

            string hash = "";
            string baseUri = "";
            if (uri.ReinterpretAs<string>().Length > 0)
            {
                int indexOfQuery = ((string)uri).IndexOf("?");
                int indexOfHash = ((string)uri).IndexOf("#");

                int indexOfEnd = Math.Min(indexOfQuery < 0 ? ((string)uri).Length : indexOfQuery,
                    indexOfHash < 0 ? ((string)uri).Length : indexOfHash);
                baseUri = ((string)uri).Substr(0, indexOfEnd);
                hash = indexOfHash < 0 ? "" : ((string)uri).Substr(indexOfHash);
            }

            return baseUri + newQueryString + hash;
        }

        /// <summary>
        /// We can find that some of our bools are actually strings, like "false" or "true". c# won't catch this
        /// and makes us write a silly method to convert our bool to another bool.
        /// </summary>
        /// <param name="value">The bool that needs sanatized</param>
        /// <returns>Null/undefined if the value passed is null. Otherse it will pass back a parsed bool</returns>
        public static bool SanatizeBoolean(bool value)
        {
            if (Script.IsNullOrUndefined(value))
            {
                return value;
            }

            return value.ToString().ToLower() == "true";
        }

        /// <summary>
        /// Disposes the given object if non-null.
        /// </summary>
        /// <returns>returns null</returns>
        /// <typeparam name="T">The type of the class to dispose</typeparam>
        [IncludeGenericArguments(false)]
        public static T Dispose<T>(T d) where T : class, IDisposable
        {
            if (Script.IsValue(d))
            {
                d.Dispose();
            }
            return null;
        }

        /// <summary>
        /// Disposes a list of disposables.
        /// </summary>
        /// <returns>returns null</returns>
        /// <typeparam name="T">The type of the class to dispose</typeparam>
        [IncludeGenericArguments(false)]
        public static List<T> Dispose<T>(List<T> d)
            where T : IDisposable
        {
            if (Script.IsValue(d))
            {
                foreach (var v in d)
                {
                    if (Script.IsValue(v))
                    {
                        v.Dispose();
                    }
                }
                d.Clear();
            }
            return null;
        }

        /// <summary>
        /// Safely calls Window.ClearTimeout for the provided handle, which was previously created using Window.SetTimeout.
        /// Returns null so that you can clear the timeout and assign in one statement, e.g.:
        /// myTimeout = MiscUtil.ClearTimeout(myTimeout);
        /// </summary>
        public static int? ClearTimeout(int? handle)
        {
            if (handle.HasValue)
            {
                Window.ClearTimeout(handle.Value);
            }
            return null;
        }

        public static int? ClearInterval(int? handle)
        {
            if (handle.HasValue)
            {
                Window.ClearInterval(handle.Value);
            }
            return null;
        }

        /// <summary>
        /// Returns a deep clone of the given object
        /// </summary>
        public static object CloneObject(object src)
        {
            string objStr = Json.Stringify(src, (k, v) =>
            {
                /*
                 * This is necessary because the runtime HUTT serializer sends over data
                 * in associative containers and javascript does not have a definition for handling them.
                 * This code snippet forces them to be interpreted as standard arrays for the purpose
                 * of serialization.
                 */
#if REMOVE_WHEN_SUPPORTED_IN_DESALT
                if (v is Uint32Array || v is Int32Array || v is Float64Array || v is Float32Array)
                {
                    // cannot call slice on v directly because typed arrays do not support slice in IE
                    // below should compile to `Array.prototype.slice.call(v);`
                    return ((dynamic)typeof(JsArray<object>).Prototype).slice.call(v);
                }
#endif
                return v;
            });

            return Json.Parse(objStr);
        }
    }
}
